src/Features/Core/Portable/CodeCleanup/AbstractCodeCleanupService.cs

```diff
@@ -136,7 +156,92 @@ protected AbstractCodeCleanupService(ICodeFixService codeFixService)
             return document;
         }
 
+        private async Task<Document> ApplyThirdPartyCodeFixesAsync(Document document, IProgressTracker progressTracker, CodeActionOptions options, CancellationToken cancellationToken)
+        {
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var range = new TextSpan(0, tree.Length);
+
+            // Compute diagnostics for everything that is not an IDE analyzer
+            var diagnostics = (await _diagnosticService.GetDiagnosticsForSpanAsync(document, range,
+                shouldIncludeDiagnostic: static diagnosticId => !(IDEDiagnosticIdToOptionMappingHelper.IsKnownIDEDiagnosticId(diagnosticId) || IsCompilerDiagnostic(diagnosticId)),
+                includeSuppressedDiagnostics: false,
+                cancellationToken: cancellationToken).ConfigureAwait(false));
+
+            // ensure more than just known diagnostics were returned
+            if (!diagnostics.Any())
+            {
+                return document;
+            }
+
+            var distinctDiagnostics = diagnostics.SelectAsArray(static d => d.Id).Distinct();
+
+            // ensure progress is reported for each third party code fix
+            progressTracker.AddItems(distinctDiagnostics.Length);
+            foreach (var diagnosticId in distinctDiagnostics)
+            {
+                using (progressTracker.ItemCompletedScope(diagnosticId))
+                {
+                    // Apply codefixes for diagnostics with a severity of warning or higher
+                    var updatedDocument = await _codeFixService.ApplyCodeFixesForSpecificDiagnosticIdAsync(
+                        document, diagnosticId, DiagnosticSeverity.Warning, progressTracker, options, cancellationToken).ConfigureAwait(false);
```

Comment:
how does this work that you're applying fixes for particular documents, but then mutating the document in place as you go?   is it recomputing diagnostics after every edit?

This will prevent any sort of compilation/semantic-model reuse.  Instead of this, can we instead just take the doc and fork it N ways for all the fixes?  then collect the changes for that particular doc, and then merge them together to make the final change?

Reply:
i'm not sure what you mean by "will succeed".  I would def not do the thing the batch fixer does where it just does things and breaks on multiple thigns writing to the same location.  But i would just have a pick-who-wins strategy and use that.  

Regardless, i'm really not comfortable with the convos happening that involve a long chain of N compilations in a row.  We've already worked hard for years to not have that because of hte debilitating perf cliff we get with this.  This is especially exacerbated now that we have generators (which will have to run after every stage of this). 

I get that this feature has value.  But there are real reasons we don't go this route (driven entirely by those aforementioned cliffs).  Even if something is high value, if it's going to involve processing like this, then we need to designa nd have a perf-acceptable approach to it up front.  This is the same thing that came up with the "fix all naming issues".  Conceptually it could function as 'run each fix after the other', but this will just fall off to badness *very* quickly.

Reply:
I'll look into updating the heuristics here beyond the very simple model we have today. I agree with the goal: if we have 10 fixes we don't want to fork 10 times and have 10 new compilations/semantic models be used.